### ISSUE TITLE
Script Effect

### ISSUE THREAD

**Issue by mkarg** [OpenShot/libopenshot#362]
**Describe the new feature:**
The set of effects in OpenShot is limited and statically compiled. This makes it hard for created people to easily (and in particular quickly) extend the abilities when working on a project.

*Script Filter* would solve that allowing the user to type in a script into a property of that filter providing the actual algorithm in an easy to learn scripting language. The script is stored within the project and can simply be copy-paste'ed to other projects.

**Describe the solution you'd like:**
Add *Script Filter* to OpenShot, having a free-text property "Script". In a first step, JavaScript could be a feasible start. In later versions, there could be a public interface allowing to extend OpenShot by custom scripting engines (like Java, Groovy, etc.) to be chosen by another property "Engine".

**Describe alternatives you have considered:**
* Custom effect
* Public interface for *external* effect libraries (.so/.dll).

**Comment by ferdnyc** [comment]
The barriers to doing this, right now, aren't so much about the scripting of effects, they're more a basic question of having _access_ to the data, in order to manipulate it.

Currently effects are, you're right, a completely closed system. (Though OpenShot/libopenshot#345 shows real promise in perhaps eventually changing that, _somewhat_.) That's in large part because of the nature of how data is manipulated within the OpenShot pipeline. Each frame of video is supplied to the effect code as an `openshot::Frame`, a custom object implemented in the libopenshot C++ code that encapsulates the frame image itself and various metadata. From the `openshot::Frame`, an `openshot::EffectBase`-derived class is able to retrieve a Qt `QImage` containing the actual video frame, which it can then manipulate (at a _very_ low level) before handing it back to the `openshot::Timeline` coordinating the processing.

It's not a setup that lends itself well to scripting or enduser-supplied manipulations, unfortunately, and despite being implemented in compiled, optimized C++ you may have noticed that currently OpenShot performance is... "not _great_," especially when attempting realtime preview. So, I'm somewhat concerned about the impact that anything like this might have on that performance, though I suppose there's an argument to be made that it can't really get _that_ much worse.

I'm going to transfer this issue over to the libopenshot repo, as any sort of implementation of this feature would have to start over there. But realistically, there are probably more fundamental improvements that would have to happen first, in order for this to be viable. Still, the _idea_ of it is sound, and I agree it would be extremely nice to have.

**Comment by jeffski** [comment]
Would certainly be a nice feature. Some more built in effects would still be great at this stage too.

**Comment by mkarg** [comment]
I think nobody would expect a **scripted** effect to perform *fast*. It serves a different purpose:
- You urgently need a non-existent effect ASAP and cannot wait until someone added it to a custom-compiled OpenShot variant or even the next official OpenShot release.
- It speeds up experimenting turnaround time with creative algorithms by far as you just can change the script and see the effect _live_, without the need to go through a build cycle of libopenshot.
- Unless we have effect **plugins** support in OpenShot, this is at least one way to get custom filters: If you need a custom effect it is better to have a *slow* one than *none*.
- Custom filters can by implemented by *non-programmers*, as scripting would be limited and easy to learn, compared to compiled languages.

In fact, as Python is a scripting language, maybe we could simply start with supporting Python before thinking in other languages: The script effect could be called "PythonEffect" then, and it would just forward the current frame as QImage to the python script, and the python script would just return the modified QImage. For creative experiments, that's all you need.

**Comment by ferdnyc** [comment]
> I think nobody would expect a **scripted** effect to perform _fast_.

Oh, you'd be _amazed_ the things people will expect. :wink:  (And the more outrageous the expectation, the more rude and demanding the request, which is the really mind-blowing thing. ...Present company excepted, of course!)

But, really my point wasn't that scripting would make things slow, it's that OpenShot itself is not set up to handle slowness.

Slow filters/effects would be fine, if you had the ability to pre-render chunks of preview so they could then be played back at something approaching a reasonable speed. But currently, the code actually explicitly _prevents_  the preview cache from getting more than <code>#CPUs</code> frames ahead of the playhead. So everything _has_ to be renderable in real time (even though it already very clearly isn't), or it can't play back with any sort of tolerable performance. That's one of the first things that really needs to change, I think.

(And none of this affects libopenshot itself, of course, which is another reason I moved this here. The possibility exists for libopenshot to support this, even if it's not usable in realtime from OpenShot itself. ...Using it from libopenshot _somewhat_ undermines Python as the most-obvious implementation choice, since it's not "already there". Still not impossible, though, as it does still have Python bindings already. libopenshot would just need either an embedded Python runtime, or an external Python helper.

**Comment by ferdnyc** [comment]
> The script effect could be called "PythonEffect" then, and it would just forward the current frame as QImage to the python script, and the python script would just return the modified QImage.

Oh, right — I'm afraid that may not work, exactly. (Or at least, it wouldn't be _quite_ that easy.) QImage data is normally passed around inside libopenshot using C++ smart pointers, so it's only accessible from within threads running in the same process. Forwarding the frames out to Python, or any other scripting language, would involve having to either write every frame to disk, as a way of handing it off to the scripting tool (and then reading the modified file back in once it's processed), or finding a way to serialize the image data for IPC purposes. Which, when we're talking about multi-megabyte frame images, could get squirrely.

OpenShot's preview is implemented inside _libopenshot_ (as a Qt widget that gets loaded in to the PyQt5 UI) is for exactly that reason: The Python side of things never has access to the actual frame image data at all, it just hosts a widget embedded from the C++ code. (Even thumbnails get written to disk by libopenshot, so they can be loaded by both the Python UI and the JS timeline.)

**Comment by ferdnyc** [comment]
(Again, I'm not saying it's not a solvable problem, and if anyone has any clever techniques for efficiently sharing large amounts of image data among code implemented in a heterogeneous set of languages, I'm all ears. This is not my area of expertise by a long shot, so the fact that I'm not aware of something is absolutely no evidence that it doesn't exist.)

**Comment by mkarg** [comment]
Thanks for sharing your insight.

I have no experience with Python actually, but I'm working with Java for more than two decades. On the JVM it is no problem at all to pass image data in-memory between a Java thread and a C++ thead, there even are standard APIs for doing that at high speed (i. e. without pixel-per-pixel copy loops). So I just assumed it would be as-easy in Python. Maybe I should set set up an lab environment to experiment with the actualy possibilites provided by Python.

Regarding preview cache: That would be a great extension anyways. ;-)

**Comment by ferdnyc** [comment]
> Regarding preview cache: That would be a great extension anyways. ;-)

Agreed, for about a billion reasons.

**Comment by ferdnyc** [comment]
> On the JVM it is no problem at all to pass image data in-memory between a Java thread and a C++ thead, there even are standard APIs for doing that at high speed (i. e. without pixel-per-pixel copy loops). So I just assumed it would be as-easy in Python.

I may even be overstating the _difficulty_ — I'm quite certain (despite what I said earlier) that it's **possible** to pass data up from Qt to Python (PyQt5 essentially already handles that). The question is whether it can be done efficiently, without either (a) forcing copies of all the objects or (b) triggering threadlock. Especially since libopenshot's frame-processing is multithreaded, and Python has the notorious [GIL](https://wiki.python.org/moin/GlobalInterpreterLock). Riverbank have always sort of been walking a tightrope, in integrating Qt's event loop with Python, a language where its own internals are explicitly _non_-thread-safe.
